crates/chia-consensus/src/gen/messages.rs

```diff
@@ -0,0 +1,271 @@
+pub const PARENT: u8 = 0b100;
+pub const PUZZLE: u8 = 0b010;
+pub const AMOUNT: u8 = 0b001;
+pub const PUZZLEAMOUNT: u8 = 0b011;
+pub const PARENTAMOUNT: u8 = 0b101;
+pub const PARENTPUZZLE: u8 = 0b110;
+pub const COINID: u8 = 0b111;
+
+#[derive(Debug)]
+pub enum SpendId {
+    OwnedCoinId(Arc<Bytes32>),
+    CoinId(NodePtr),
+    Parent(NodePtr),
+    Puzzle(NodePtr),
+    Amount(u64),
+    PuzzleAmount(NodePtr, u64),
+    ParentAmount(NodePtr, u64),
+    ParentPuzzle(NodePtr, NodePtr),
+    None,
+}
+
+impl SpendId {
+    // args is an in-out parameter. It's updated to point to then next argument
+    pub fn parse(a: &Allocator, args: &mut NodePtr, mode: u8) -> Result<SpendId, ValidationErr> {
+        // we have a special case for when all three mode flags are set. That means
+        // we're committing to parent, puzzle and amount. In this case you just
+        // specify the coin ID
+        if mode == COINID {
+            let coinid = sanitize_hash(a, first(a, *args)?, 32, ErrorCode::InvalidCoinId)?;
+            *args = rest(a, *args)?;
+            return Ok(Self::CoinId(coinid));
+        }
+
+        let parent = if (mode & PARENT) != 0 {
+            let parent = sanitize_hash(a, first(a, *args)?, 32, ErrorCode::InvalidParentId)?;
+            *args = rest(a, *args)?;
+            parent
+        } else {
+            NodePtr::NIL
+        };
+
+        let puzzle = if (mode & PUZZLE) != 0 {
+            let puzzle = sanitize_hash(a, first(a, *args)?, 32, ErrorCode::InvalidPuzzleHash)?;
+            *args = rest(a, *args)?;
+            puzzle
+        } else {
+            NodePtr::NIL
+        };
+
+        let amount = if (mode & AMOUNT) != 0 {
+            let amount = match sanitize_uint(a, first(a, *args)?, 8, ErrorCode::InvalidCoinAmount)?
+            {
+                SanitizedUint::PositiveOverflow => {
+                    return Err(ValidationErr(*args, ErrorCode::AmountExceedsMaximum));
+                }
+                SanitizedUint::NegativeOverflow => {
+                    return Err(ValidationErr(*args, ErrorCode::NegativeAmount));
+                }
+                SanitizedUint::Ok(amount) => amount,
+            };
+            *args = rest(a, *args)?;
+            amount
+        } else {
+            0
+        };
+
+        match mode {
+            PARENT => Ok(Self::Parent(parent)),
+            PUZZLE => Ok(Self::Puzzle(puzzle)),
+            AMOUNT => Ok(Self::Amount(amount)),
+            PARENTPUZZLE => Ok(Self::ParentPuzzle(parent, puzzle)),
+            PARENTAMOUNT => Ok(Self::ParentAmount(parent, amount)),
+            PUZZLEAMOUNT => Ok(Self::PuzzleAmount(puzzle, amount)),
+            0 => Ok(Self::None),
+            _ => Err(ValidationErr(*args, ErrorCode::InvalidMessageMode)),
+        }
+    }
+
+    pub fn from_self(
+        mode: u8,
+        parent: NodePtr,
+        puzzle: NodePtr,
+        amount: u64,
+        coin_id: &Arc<Bytes32>,
```

Comment:
Again here, I think `coin_id` should just be `Bytes32`. A reference to an `Arc` is unnecessary for this imo.

Reply:
in the stress tests I have, avoiding copying this gives a material speed-up. Granted, I don't have stress tests for these conditions specifically, but there are other stress tests in `sr/gen/test_generators.rs` for `CREATE_COIN` and `CREATE_COIN_ANNOUNCEMENT`

Reply:
I'm surprised this is faster, I would think copying 32 bytes to be quite fast. I've been writing `[u8; 32]` instead of `&[u8; 32]` for this reason - but I'll keep it in mind.

Reply:
as long as everything stays in L1 (or maybe even L2) I would share your assumption. However, the stress tests deliberately causes memory pressure to make the whole thing be evicted into L3 and even main RAM. Then you're looking at 2 orders of magnitude longer delays